File: server/src/retrieval/reranker.js
import { cosineSimilarity } from '../rag.js';

// No cross-encoder model is available locally (CPU-only Ollama box), so this
// reranks using Maximal Marginal Relevance over the embeddings already
// computed during retrieval: balances relevance to the query against
// redundancy with chunks already selected, instead of just taking the top-N
// by raw score (which tends to return 5 near-duplicate chunks from the same
// paragraph).
const MMR_LAMBDA = 0.7; // weight toward relevance vs diversity

export async function rerank(queryVector, candidates, topK = 8) {
  if (candidates.length === 0) return [];

  const pool = candidates.map(c => ({
    ...c,
    relevance: c.embedding ? cosineSimilarity(queryVector, c.embedding) : (c.hybridScore || 0)
  }));

  const selected = [];
  const remaining = [...pool];

  while (selected.length < topK && remaining.length > 0) {
    let bestIdx = 0;
    let bestScore = -Infinity;

    for (let i = 0; i < remaining.length; i++) {
      const candidate = remaining[i];
      const maxSimToSelected = selected.length === 0
        ? 0
        : Math.max(...selected.map(s => (
          candidate.embedding && s.embedding ? cosineSimilarity(candidate.embedding, s.embedding) : 0
        )));
      const mmrScore = MMR_LAMBDA * candidate.relevance - (1 - MMR_LAMBDA) * maxSimToSelected;
      if (mmrScore > bestScore) {
        bestScore = mmrScore;
        bestIdx = i;
      }
    }

    const [picked] = remaining.splice(bestIdx, 1);
    selected.push({ ...picked, rerankScore: bestScore });
  }

  return selected;
}

File: server/app/retrieval/reranker.py
import re
from typing import List, Optional
from ..rag import cosine_similarity

MMR_LAMBDA = 0.7  # weight toward relevance vs diversity


def _token_jaccard(text_a: str, text_b: str) -> float:
    set_a = set(re.findall(r"\w+", text_a.lower()))
    set_b = set(re.findall(r"\w+", text_b.lower()))
    if not set_a or not set_b:
        return 0.0
    return len(set_a & set_b) / len(set_a | set_b)


async def rerank(query_vector: Optional[list[float]], candidates: List[dict], top_k: int = 8) -> List[dict]:
    """Lightweight MMR reranker combining semantic/composite relevance and inter-chunk
    diversity. Operates efficiently without loading expensive local cross-encoders."""
    if not candidates:
        return []

    pool = []
    for c in candidates:
        # If composite_score is present from freshness_scorer, use it; otherwise compute cosine
        if "composite_score" in c:
            rel = float(c["composite_score"])
        elif query_vector and c.get("embedding"):
            rel = float(cosine_similarity(query_vector, c["embedding"]))
        else:
            rel = float(c.get("hybridScore", 0.0) * 10.0 or c.get("similarity", 0.0))

        pool.append({**c, "relevance": rel})

    selected: List[dict] = []
    remaining = list(pool)

    while len(selected) < top_k and remaining:
        best_idx = 0
        best_score = float("-inf")

        for i, candidate in enumerate(remaining):
            if not selected:
                max_sim = 0.0
            else:
                max_sim = 0.0
                for s in selected:
                    if candidate.get("embedding") and s.get("embedding"):
                        sim = cosine_similarity(candidate["embedding"], s["embedding"])
                    else:
                        sim = _token_jaccard(candidate.get("content", ""), s.get("content", ""))
                    if sim > max_sim:
                        max_sim = sim

            mmr_score = MMR_LAMBDA * candidate["relevance"] - (1.0 - MMR_LAMBDA) * max_sim
            if mmr_score > best_score:
                best_score = mmr_score
                best_idx = i

        picked = remaining.pop(best_idx)
        selected.append({**picked, "rerankScore": round(best_score, 4)})

    return selected
